include/alpaka/atomic/AtomicStdLibLock.hpp

```diff
@@ -81,22 +83,40 @@ namespace alpaka
         template<typename TOp, typename T, typename THierarchy, size_t THashTableSize>
         struct AtomicOp<TOp, AtomicStdLibLock<THashTableSize>, T, THierarchy>
         {
+            static constexpr bool IsLockFree
+                = std::is_trivially_copyable_v<T> && boost::atomic_ref<T>::required_alignment <= alignof(T)
+                && boost::atomic_ref<T>::is_always_lock_free;
             ALPAKA_FN_HOST static auto atomicOp(
                 AtomicStdLibLock<THashTableSize> const& atomic,
                 T* const addr,
                 T const& value) -> T
             {
-                std::lock_guard<std::mutex> lock(atomic.getMutex(addr));
-                return TOp()(addr, value);
+                if constexpr(IsLockFree)
+                {
+                    return TOp()(addr, value);
+                }
+                else
+                {
+                    std::lock_guard<std::mutex> lock(atomic.getMutex(addr));
+                    return TOp()(addr, value);
+                }
             }
```

Comment:
I wonder if we could always use `boost::atomic_ref` as implementation. I believe the cases where `boost::atomic_ref`  would not work, are probably also not going to work in CUDA.

Reply:
The CUDA backend has its own atomic implementation. This one is used by `AccCpuSerial`, `AccCpuTbbBlocks`, `AccCpuFibers`, `AccCpuThreads` and at between grids by `AccCpuOmp2Blocks` and `AccCpuOmp2Threads`. it does not affect other backends.

Reply:
Sure! But what I meant was that the implemented `if constexpr` here and inside the functors avoid `boost::atomic_ref` for `T`s which are not copyable or badly aligned. And I argued that if `T` is something that is not copyable or badly aligned, that will likely also not work with the CUDA atomics, so we could just use the boost implementation for all `T`s on the backends you listed.

Reply:
According to the [CUDA documentation](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#atomic-functions)
> An atomic function performs a read-modify-write atomic operation on one 32-bit or 64-bit word residing in global or shared memory.

According to the [Boost documentation](https://www.boost.org/doc/libs/1_77_0/doc/html/atomic/interface.html#atomic.interface.interface_atomic_ref)
> The referenced object must have alignment not less than indicated by the `boost::atomic_ref<T>::required_alignment` constant. That constant may be larger than the natural alignment of type `T`.

In our test I did not find any cases where this check would fail, but I don't know what are the internals of the Boost implementation.

Reply:
I do not not think that the fallback lock version hurts us very much so I would be fine if we keep it.